### ISSUE TITLE
Layout support

### ISSUE THREAD

**Issue by psiha** [jfalcou/kiwaku#8]
A classic ML example are the NHWC (batch index, height, width, channel - channel interleaved layout) and/vs NCHW (channel separated layout) layouts.
Consider some sort of builtin support (generic layout definition mechanism and mapping and converting between compatible layouts).

**Comment by psiha** [comment]
So, in the meantime, I implemented this on top of kiwaku (and it is just getting into production), I got permission to share the core of that functionality - so we can work on this further together/bring it into kiwaku:

[Dimension.hpp](https://gist.github.com/psiha/8b72d7f304f370caf0343187da75608a)
- simple header, holds the basic machinery for defining a 'dimension' (i.e. a typed axis)
- predefines N, H, W, C

[Dimensions.hpp](https://gist.github.com/psiha/f57baa3e423288f8ca8a961eda3c92c7)
- complex header, something like your shape.hpp, in fact, it hosts the Dimensions<> class template that wraps your shape<>
- Dimensions<> 'holds'/remembers the type information of the specified dimensions/axes and delegates to shape for storage, access, indexing (although this it has to 'transform'/'correct' due to https://github.com/jfalcou/kiwaku/issues/39)
- so basically you can say:
Dimensions< N[ 1 ], H, W, C[ 32 ] > dims0;
dims0.get< C >();
Dimensions< H, W, C > dims1{ dims0 }; // there are complex and somewhat documented conversion rules
auto dims3{ dims1.pushFront< N[ 3 ] >() };
...

on top of these I built NDArrays (like your arrays) but we can get to those later...

...so if you can take a look and see how it maps to your vision...we can also discuss/go through it in a video call..

**Comment by jfalcou** [comment]
Sounds interesting. I'll dig into it. We can do a zoom or w/e after the 21, pretty busy atm :D

**Comment by jfalcou** [comment]
I like the named axis thing, this is maybe a good interface for the storage order support, at least better than a random permutation-like function.

**Comment by jfalcou** [comment]
#76 implements some of those ideas.
Feedback welcome.

**Comment by psiha** [comment]
> @psiha let me now when you are available next week or whenever for a quick visio on Discord to discuss that. I have a bunch of design idea base don your code that I want to bounce off before comittign to code them.

sorry...as you may have noticed 🙈 i had to divert my attention just as "we were gettin' somewhere" here :/
anyways kiwaku is on my docket again...going through the changes right now...next week i'll be mostly free to catch up, discuss and brainstorm

**Comment by jfalcou** [comment]
LEt's do that. Cause we also support the "named" axis thing and I think we are close to what you are doing.
